src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -2858,29 +2924,35 @@ public function get_modifiable_text(): string {
 			'STYLE' === $tag_name ||
 			'XMP' === $tag_name
 		) {
-			return $text;
+			return str_replace( "\x00", "\u{FFFD}", $text );
 		}
 
 		$decoded = WP_HTML_Decoder::decode_text_node( $text );
 
 		/*
-		 * TEXTAREA skips a leading newline, but this newline may appear not only as the
-		 * literal character `\n`, but also as a character reference, such as in the
-		 * following markup: `<textarea>&#x0a;Content</textarea>`.
+		 * Skip the first line feed after LISTING, PRE, and TEXTAREA opening tags.
 		 *
-		 * For these cases it's important to first decode the text content before checking
-		 * for a leading newline and removing it.
+		 * Note that this first newline may come in the form of a character
+		 * reference, such as `&#x0a;`, and so it's important to perform
+		 * this transformation only after decoding the raw text content.
 		 */
 		if (
-			self::STATE_MATCHED_TAG === $this->parser_state &&
-			'TEXTAREA' === $tag_name &&
-			strlen( $decoded ) > 0 &&
-			"\n" === $decoded[0]
+			( "\n" === ( $decoded[0] ?? '' ) ) &&
+			( ( $this->skip_next_linefeed && '#text' === $tag_name ) || 'TEXTAREA' === $tag_name )
 		) {
-			return substr( $decoded, 1 );
+			$decoded = substr( $decoded, 1 );
 		}
+		$this->skip_next_linefeed = false;
```

Comment:
it's hard to change when advancing to the next token, because that wipes out the flag when moving from LISTING and PRE to `#text`.

I'll explore setting `last_node_name` or something like this. it's a good catch, thank you.

Reply:
in [1323448](https://github.com/WordPress/wordpress-develop/pull/6972/commits/1323448ce22f6c8811caf7f50858c06df4674309) I've changed this to `$this->skip_newline_at` as a way of declaratively detecting where to ignore the newline. this is more resilient, but I still have concerns about seeking and updating the document.

I'll try to figure out seeking, but as it stands we don't expose a way to change a PRE or LISTING tag into something else, so I will not worry too much about fixing those cases until we allow getting into them.

Reply:
for now I've simply listed the behavior after seeking as a limitation and warned to seek to the PRE or LISTING, not to the `#text` node. I don't know how to fix this without introducing unseen overhead in the Tag Processor.

do you think this is reasonable enough?